src/Access/AccessControl.cpp

```diff
@@ -62,6 +63,52 @@ namespace
                 users_config_path, config_path);
         }
     }
+
+    PreformattedMessage getAuthFailedExceptionMessage(bool login_failure_quota_exceeded, const Credentials & credentials)
+    {
+        WriteBufferFromOwnString message;
+        message << credentials.getUserName() << ": Authentication failed: ";
+
+        if (!login_failure_quota_exceeded)
+            message << "password is incorrect, or there is no user with such name.";
+        else
+            message << "login failure quota exceeded.";
+
+        /// Better exception message for usability.
+        /// It is typical when users install ClickHouse, type some password and instantly forget it.
+        if (credentials.getUserName().empty() || credentials.getUserName() == "default")
+            message << "\n\n"
+                    << "If you have installed ClickHouse and forgot password you can reset it in the configuration file.\n"
+                    << "The password for default user is typically located at /etc/clickhouse-server/users.d/default-password.xml\n"
+                    << "and deleting this file will reset the password.\n"
+                    << "See also /etc/clickhouse-server/users.xml on the server where ClickHouse is installed.\n\n";
+
+        /// We use the same message for all authentication failures because we don't want to give away any unnecessary information for security reasons,
+        /// only the log will show the exact reason.
+
+        if (!login_failure_quota_exceeded)
+            return PreformattedMessage{message.str(), "{}: Authentication failed: password is incorrect, or there is no user with such name.{}"};
+        else
+            return PreformattedMessage{message.str(), "{}: login failure quota exceeded.{}"};
+    }
+
+    bool addLoginFailureToQuota(std::shared_ptr<const EnabledQuota> & authentication_quota)
+    {
+        bool quota_exceed = false;
+        if (authentication_quota)
+        {
+            try
+            {
+                authentication_quota->used(QuotaType::FAILED_SEQUENTIAL_AUTHENTICATIONS, 1, true);
+            }
+            catch (const Exception & /* e */)
+            {
+                quota_exceed = true;
```

Comment:
I think it's better to use a different error code and error message if the quota is exceeded, not just usual `AUTHENTICATION_FAILED`. Just for usability - it seems to be useful to know that an attempt has failed due to too many attempts (not because the password was incorrect).

Reply:
Done! Code is simplified.